### PR TITLE
[Merged by Bors] - feat: the multiGoal linter

### PR DESCRIPTION
**Pull request by adomani**
A linter that makes sure that, when multiple goals are present, they are enclosed in `·`s.

Adaptations (the order is chronological, there should be no dependency among them):
* #12338,
* #12361,
* #12372,
* #12560,
* #12834,
* #12381,
* #12908,
* #14939,

plus many many more that this comment is too small to contain.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: test/Multigoal.lean [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+#guard_msgs(drop warning) in
+set_option linter.multiGoal true in
+/--
+warning: 'exact .intro' leaves 1 goal 'Lean.Parser.Tactic.exact'
+note: this linter can be disabled with `set_option linter.multiGoal false`
+-/
+#guard_msgs in
+example : True := by
+  by_cases 0 = 0
```
Comment (grunweg) [inline test/Multigoal.lean:25]: Trying to articulate when the linter errors gave me an idea about how to phrase the error message.
- I could be nice to say "`by_cases 0 = 0` produced two goals", but that might be hard in general
- Something that *is* feasible is to change the wording of the error, something like
```
warning: there are 2 unclosed goals before 'exact .intro', and one stays unsolved
please focus on the current goal, for instance using `.` (typed as "\.")

File: test/Multigoal.lean [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  exact .intro
+  exact .intro
+
+#guard_msgs(drop warning) in
+set_option linter.multiGoal true in
+/--
+warning: 'assumption' leaves 1 goal 'Lean.Parser.Tactic.assumption'
+note: this linter can be disabled with `set_option linter.multiGoal false`
+-/
+-- the linter keeps linting after ignoring a `conv`, `conv_lhs`, `conv_rhs`
```
Comment (grunweg) [inline test/Multigoal.lean:35]:
```suggestion
-- The linter ignores `conv`, `conv_lhs` and `conv_rhs`, but keeps linting afterwards.

File: test/Multigoal.lean [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  iterate left; decide
+  repeat' left; decide
+  refine Or.inl ⟨?_, ?_⟩
+  rfl
+  assumption
+
+#guard_msgs(drop warning) in
+set_option linter.multiGoal true in
+/--
+warning: 'rfl' leaves 2 goals 'Lean.Parser.Tactic.tacticRfl'
```
Comment (grunweg) [inline test/Multigoal.lean:72]: Added this test, to clarify which lines warn, and to show that several warnings also work.

File: test/Multigoal.lean [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  trivial
+  rfl
+
+example (p : Bool) : 0 = 0 := by
+  cases p
+  case' false => rfl
+  case' true => rfl
+
+#guard_msgs in
+-- `assumption'` is allowed, as it is useful precisely when there are multiple active goals
```
Comment (grunweg) [inline test/Multigoal.lean:91]:
```suggestion
-- `assumption'` is allowed, as it is useful precisely when there are multiple active goals.

File: test/Multigoal.lean [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+example (p : Bool) (f : False) {h : 0 = 0} : 0 = 0 ∧ 0 = 1 := by
+  cases p <;>
+  constructor
+  assumption'
+  any_goals cases f
+
+#guard_msgs in
+-- `focus` is ignored
```
Comment (grunweg) [inline test/Multigoal.lean:99]: What do you think about "... is allowed as a proof scoping tactic"?

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+* Should the linter flag unnecessary scoping as well?
+  For instance, should
+  ```lean
+  example : True := by
+    · · exact .intro
+  ```
+  raise a warning?
+* Custom support for "accumulating side-goals", so that once they are all in scope
+  they can be solved in bulk via `all_goals` or a similar tactic.
+* In development, `on_goal` has been partly used as a way of silencing the linter
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:30]: Is this item still current? My understanding is that nobody complained much about this.

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+The "multiGoal" linter emits a warning where there is more than a single goal in scope.
+
+There are a few tactics that are intended to work specifically in such a situation and the linter
+ignores them. Otherwise, whenever a tactic leaves multiple goals, the linter will emit a warning,
+unless some form of "focusing" tactic is used.
+Typically, the focusing is achieved by the `cdot`: `·`, but, e.g., `focus` or `on_goal x` also
+serve a similar purpose.
+
+TODO:
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:20]:
```suggestion
## TODO

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Lean.Elab.Command
+import Lean.Linter.Util
+
+/-!
+#  The "multiGoal" linter
+
+The "multiGoal" linter emits a warning where there is more than a single goal in scope.
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:12]: Re-re-thinking this wording, I am not completely happy: tactics creating side goals are fine. The point is when there are multiple active goals, but a tactic only acts on one of them. What do you think about something like the following?

```
The "multiGoal" linter emits a warning when there is more than a single goal in scope, but a tactic only acts on one of them.

As mandated by the mathlib style guide, when new goals arise as side conditions or steps,
they should be focused. It is common for tactics to split one goal into several steps, or to leave side conditions to be checked. In these cases, goals should be focused until there is only one active goal at a time. Typically, this focusing is achieved by the `cdot`: `·`, but, the `focus` or `on_goal` tactics (for example) serve a similar purpose.

The only exception to this rule are tactic combinators like `all_goals`, intended to run one or several tactics on several goals at once.
```

Comment (grunweg) [reply Mathlib/Tactic/Linter/Multigoal.lean:12]: I am in fact not fully sure if the first sentence is correct: would the following situation error?
-- one goal before
tactic1 -- two goals now
tactic2 -- closes both

In any case, can you add a test for something like this? My last adaptation PR contained one.

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+namespace multiGoal
+
+/-- These are the `SyntaxNodeKind`s of tactics that we allow to have "inactive" goals.
+Reasons for admitting a kind in `exclusions` include
+* the tactic is reordering the goals, e.g. `swap`, `rotate_left`, ...;
+* the tactic is structuring a proof, e.g. `skip`, `<;>`, ...;
+* the tactic is creating new goals, e.g. `constructor`, `cases`, `induction`, ....
+Tactic combinators like `repeat` or `try` are a mix of both.
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:52]: Thinking closer: I think it's clearer to delete this sentence, and move tactic combinators to the next list.

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+  |>.insert `Mathlib.Tactic.Conv.convLHS
+  |>.insert `Mathlib.Tactic.Conv.convRHS
+  |>.insert ``Lean.Parser.Tactic.first
+  |>.insert ``Lean.Parser.Tactic.repeat'
+  |>.insert ``Lean.Parser.Tactic.tacticIterate____
+  |>.insert ``Lean.Parser.Tactic.anyGoals
+  |>.insert ``Lean.Parser.Tactic.allGoals
+  |>.insert ``Lean.Parser.Tactic.focus
+
+/-- `getManyGoals` returns the syntax nodes where the tactic leaves at least one goal that
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:107]: In terms of style, what do you think about this rephrasing?
`getManyGoals t` returns the syntax nodes in `t` corresponding to tactic calls which
- leave at least one goal that was not present before it ran
- are not excluded through `exclusions` or `ignoreBranch`

(This needs to be adapted if `ignoreBranch` is renamed or the logic tweaked, as indicated below.)

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+was not present before it ran,
+unless its `SyntaxNodeKind` is either in `exclusions` or in `ignoreBranch`.
+-/
+partial
+def getManyGoals : InfoTree → Array (Syntax × Nat)
+  | .node k args =>
+    let kargs := (args.map getManyGoals).foldl (· ++ ·) #[]
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:114]: Does `(args.map getManyGoals).flatten` also work? This is what the `refine'` linter uses.

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    if let .ofTacticInfo i := k then
+      if ignoreBranch.contains i.stx.getKind then #[] else
+      if let  .original .. := i.stx.getHeadInfo then
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:117]: Can you comment this branch? Why do you need it?

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        let newGoals := i.goalsAfter.filter (i.goalsBefore.contains ·)
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:118]: As you wrote via DM, this uses a heuristic: you use the heuristic "meta-variables have the same name" to determine whether a tactic only modifies a goal, or creates new ones. Can you add a comment to this effect?

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+        if newGoals.length != 0 && !exclusions.contains i.stx.getKind then
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:119]: This is my only high-level comment: I agree the code does what it says, but I wonder if this is what you or I want.

Intuitively, I would rather expect the linter to fire if "goalsBefore.length > 2 and goalsAfter.length != 0", which is slightly different. (This is the wording I proposed above.) What do you think?

**Review by grunweg — COMMENTED** [review]
Thanks for writing this linter! The code is quite short, but I know that this always takes more thought than one might think. The linter does what it says and looks well-written. I have one high-level comment: intuitively, I would expect its logic to be slightly different (see below). 

I also reviewed everything else carefully: as you know by now, that entails some word-smithing suggestions and a number of minor comments. (I pushed some tweaks directly; feel free to revert or discuss anything you do not agree with.)

**Comment by grunweg** [comment]
I just fixed the latest round of issues, and split these into #16331.

File: Mathlib/Tactic/Linter/Multigoal.lean (hunk 13 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/Multigoal.lean]: Never mind, that comment applies to the flexible linter, but not this one.

Comment (grunweg) [reply Mathlib/Tactic/Linter/Multigoal.lean]: Never mind, that comment applies to the flexible linter, but not this one.

File: Mathlib/Tactic/Linter/Multigoal.lean (hunk 8 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/Multigoal.lean:11]: You are right: there is an extra exception which is a tactic that starts with more than one open goals and the tactic closes *all* goals.  I.e., the linter shuts up if no goals are left.  I added this to the description.

File: test/Multigoal.lean (hunk 1 above)
Comment (adomani) [reply test/Multigoal.lean:27]: I also updated the error message, thanks!

**Review by adomani — COMMENTED** [review]
Michael, thank you very much for the comment!  I think that this is (finally) ready again for review!

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    ``Lean.Parser.Tactic.injections,
+    ``Lean.Parser.Tactic.substVars,
+    `Batteries.Tactic.«tacticPick_goal-_»,
+    ``Lean.Parser.Tactic.case',
+    `«tactic#adaptation_note_»,
+    `tacticSleep_heartbeats_
+  ]
+
+/-- The `SyntaxNodeKind`s in `ignoreBranch` correspond to tactics that disable the linter from
+their first application until the corresponding proof branch is closed.
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:103]: Thank you; this is much clearer! I'm stumbling a bit over the term "proof branch" - I don't know a better term though. (Do you?)

File: Mathlib/Tactic/Linter/Multigoal.lean (hunk 9 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/Multigoal.lean:55]: Thank you; this helps!

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+
+/-- The `SyntaxNodeKind`s in `exclusions` correspond to tactics that the linter allows,
+even though there are multiple active goals.
+Reasons for admitting a kind in `exclusions` include
+* the tactic focuses on one goal, e.g. `·`, `focus`, `on_goal i =>`, ...;
+* the tactic is reordering the goals, e.g. `swap`, `rotate_left`, ...;
+* the tactic is structuring a proof, e.g. `skip`, `<;>`, ...;
+* the tactic is creating new goals, e.g. `constructor`, `cases`, `induction`, ....
+
+Tactic combinators like `repeat` or `try` are a mix of both.
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:55]: Should this sentence move after the "There is some overlap in scope between `ignoreBranch` and `exclusions`."? (Otherwise, what does "of both" refer to?)

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+which
+* leave at least one goal that was not present before it ran;
+* are not excluded through `exclusions` or `ignoreBranch`.
+-/
+partial
+def getManyGoals : InfoTree → Array (Syntax × Nat)
+  | .node info args =>
+    let kargs := (args.map getManyGoals).toArray.flatten
+    if let .ofTacticInfo info := info then
+      if ignoreBranch.contains info.stx.getKind then #[] else
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:132]: Very minor nit: move the "else" to the beginning of the next line? Up to you.

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+    `Mathlib.Tactic.Conv.convRHS,
+    ``Lean.Parser.Tactic.first,
+    ``Lean.Parser.Tactic.repeat',
+    ``Lean.Parser.Tactic.tacticIterate____,
+    ``Lean.Parser.Tactic.anyGoals,
+    ``Lean.Parser.Tactic.allGoals,
+    ``Lean.Parser.Tactic.focus
+  ]
+
+/-- `getManyGoals t` returns the syntax nodes of the `InfoTree` `t` corresponding to tactic calls
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:122]: Can you also mention what the returned natural number is, please?

File: lakefile.lean [hunk 19]
```diff
@@ -36,6 +36,8 @@ abbrev mathlibOnlyLinters : Array LeanOption := #[
   ⟨`linter.style.longLine, true⟩,
   ⟨`linter.style.longFile, .ofNat 1500⟩,
   ⟨`linter.style.missingEnd, true⟩,
+  ⟨`linter.style.multiGoal, true⟩,
+  ⟨`linter.style.setOption, true⟩,
```
Comment (grunweg) [inline lakefile.lean:40]: A merge conflict, I guess:
```suggestion

File: test/Multigoal.lean [hunk 20]
```diff
@@ -0,0 +1,115 @@
+import Mathlib.Tactic.Basic
+import Mathlib.Tactic.Conv
+import Mathlib.Tactic.Linter.Multigoal
+import Mathlib.Util.SleepHeartbeats
+
+-- The warning generated by `linter.style.multiGoal` is not suppressed by `#guard_msgs`,
+-- because the linter is run on `#guard_msgs` itself. This is a known issue, see e.g.
+-- https://leanprover.zulipchat.com/#narrow/stream/348111-batteries/topic/unreachableTactic.20linter.20not.20suppressed.20by.20.60.23guard_msgs.60
+-- We jump through an extra hoop here to silence the warning.
```
Comment (grunweg) [inline test/Multigoal.lean:9]: Let's get this linter in before the Lean bump :-)

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+  run := withSetOptionIn fun _stx => do
+    unless Linter.getLinterValue linter.style.multiGoal (← getOptions) do
+      return
+    if (← MonadState.get).messages.hasErrors then
+      return
+    let trees ← getInfoTrees
+    for t in trees.toArray do
+      for (s, n) in getManyGoals t do
+        Linter.logLint linter.style.multiGoal s
+          m!"There are {n+1} unclosed goals before '{s}' and at least one remaining goal after.\n\
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:154]: Not a native speaker, but: is "afterwards" better English?

File: Mathlib/Tactic/Linter/Multigoal.lean (hunk 8 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/Multigoal.lean:11]: Thanks, this definitely helps! Do you have the energy to add a test for such a situation? (Would that be interesting? My motivation is to ensure these edge cases are not changed accidentally, and adding a test is an easy way to do so. Mathlib proofs change over time.)

File: Mathlib/Tactic/Linter/Multigoal.lean [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+-/
+import Lean.Elab.Command
+
+/-!
+#  The "multiGoal" linter
+
+The "multiGoal" linter emits a warning where there is more than a single goal in scope.
+There is an exception: a tactic that closes *all* remaining goals is allowed.
+
+There are a few tactics that are intended to work specifically in such a situation and the linter
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Multigoal.lean:14]: This is getting into copy-editing territory: this description covers everything - but re-reading this again (after a few months), I find this hard to parse. There are exceptions, and exceptions of exceptions... can the same content be said more clearly and directly? Something like:

```
Suppose there are multiple goals in scope, and the next tactic call does not close them all. (This is a common situation, as tactic can split one goal into several steps, or to leave side conditions to be checked.)

There are a few tactics (such as `skip`, `skip` or the `try` combinator) that are intended to work specifically in such a situation. Otherwise, the mathlib style guide ask that goals be be focused until there is only one active goal at a time.
If this focusing does not happen, the linter emits a warning.
Typically, the focusing is achieved by the `cdot`: `·`, but, e.g., `focus` or `on_goal x` also serve a similar purpose.
```

This is longer and less concise, but (I think) clearer: less complicated sentences, less negation

**Review by grunweg — COMMENTED** [review]
Thank you for the changes (and sorry that I didn't find time to propose something myself). In particular, the new error message seems much clearer to me! I have some more comments on the comments. I think that is close to all I have to comment on this PR.

File: Mathlib/Tactic/Linter/Multigoal.lean (hunk 17 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/Multigoal.lean:136]: Moved!

File: Mathlib/Tactic/Linter/Multigoal.lean (hunk 18 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/Multigoal.lean:124]: Done!

File: Mathlib/Tactic/Linter/Multigoal.lean (hunk 21 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/Multigoal.lean]: I don't know, both sound correct to me, so I went with your version!

File: lakefile.lean (hunk 19 above)
Comment (adomani) [reply lakefile.lean]: Indeed, badly resolved!

**Review by adomani — COMMENTED** [review]
Michael, thank you for all your reviewing effort!

**Comment by kim-em** [comment]
bors d+

**Comment by adomani** [comment]
Bors merge

**Comment by kim-em** [comment]
I guess bors is case sensitive?

bors merge

**Comment by kim-em** [comment]
Nope, just didn't add the label for some reason.

**Comment by adomani** [comment]
I had checked that `bors` was running, but did not notice the missing label...
